blockchain.go

```diff
@@ -341,6 +343,51 @@ func (b *Blockchain) GetCommonAncestor(bestHeader, prevBestHeader StoredHeader)
 	}
 }
 
+// Rollback the header database to the last header before time t.
+// We shouldn't go back further than the checkpoint
+func (b *Blockchain) Rollback(t time.Time) error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	checkpoint := GetCheckpoint(b.crationDate, b.params)
+	checkPointHash := checkpoint.Header.BlockHash()
+	sh, err := b.db.GetBestHeader()
+	if err != nil {
+		return err
+	}
+	// If t is greater than the timestamp at the tip then do nothing
+	if sh.header.Timestamp.Before(t) {
+		return nil
+	}
+	// If the tip is our checkpoint then do nothing
+	checkHash := sh.header.BlockHash()
+	if checkHash.IsEqual(&checkPointHash) {
+		return nil
+	}
+	rollbackHeight := uint32(0)
+	for {
```

Comment:
We should probably put a safety limit on this. Something larger than what can happen under correct circumstances, e.g. 10 billion, but can prevent getting stuck in a endless loop that eats all the CPU.